src/boot.c

```diff
@@ -690,10 +692,29 @@ static void write_volume_label(DOS_FS * fs, char *label)
 {
     time_t now = time(NULL);
     struct tm *mtime = localtime(&now);
+    char *source_date_epoch = getenv("SOURCE_DATE_EPOCH");
     off_t offset;
     int created;
     DIR_ENT de;
 
+    if (source_date_epoch) {
+        char *endptr = NULL;
+        errno = 0;
+        now = strtoll(source_date_epoch, &endptr, 10);
```

Comment:
`now` is of type `time_t` which is 32bit on 32bit systems. `strtoll` returns `long long` value (at least 64bit) which overflow when storing into 32bit type.

Reply:
Writing this correctly is difficult. There is no TIME_T_MAX to check against, and apparently time_t can be any integer or float type. And even as integer type, it can have "padding bits", so one cannot check the max possible value based on sizeof(time_t).

However, if we assume time_t is long int, which seems to be the most common type, how about this?

```
time_t now = 0;
long long val = strtoll(...);
/* check for strtoll success */

if (sizeof(now) == sizeof(int32_t)) {
    if (val > INT32_MAX) {
        /* overflow, abort */
    } else {
        now = val;
    }
} else if (sizeof(now) >= sizeof(val)) {
    now = val;
} else {
    /* Unsupported time_t type/size, abort. */
    /* Or add build time assert about time_t type/size? */
}
```

Reply:
In your example you have a problem with signedness, when now and val differs.

In past I proposed following check for underflow/overflow:
```c
long long val;
...
time_t now = val;
if (((time_t)-1 > 0 && val < 0) || now > LLONG_MAX || (long long)now != val) {
    /* conversion failed */
}
```

If you want to be pedant and check also floating point time_t type, then you can use:
```c
if ((time_t)1 / 2 == 0) {
    /* time_t is integer */
} else {
    /* time_t is floating point
}
```

In case it is floating point, you can probably ignore problems with casting as in most cases you loose precision. Or use one of `strtof`/`strtod`/`strtold` function.

Reply:
Nice and compact. I like it! But I don't understand how ```now > LLONG_MAX``` can ever be true, after the ```time_t now = val;``` line. Care to explain?

Reply:
This is theoretical case, when time_t is 64bit unsigned type.

Basically it is safe check prior using `(long long)now`, to check that casting does not overflow. I'm doing these checks prior comparing and casting in any code.

Reply:
When building for 32-bit system, gcc prints this:

```
warning: comparison is always false due to limited range of data type [-Wtype-limits]
             if (((time_t)-1 > 0 && val < 0) || now > LLONG_MAX || (long long)now != val) {
                                                    ^
```

What's a nice way to deal with that warning?

Reply:
It is wrong/false-positive warning. Reason why warnings in production builds are turned off...
You can compile with -Wno-type-limits or add some gcc pragma which turns that warning locally (I do not remember how to control them right now).

Reply:
Or you can try to play with `__builtin_choose_expr` or with `__builtin_constant_p`. See:
https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fchoose_005fexpr
https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fconstant_005fp

Reply:
Or drop that `now > LLONG_MAX` check. Negative `val` with unsigned `time_t` is already checked by previous condition.

Reply:
Ah, even better, no platform specific code.

Since I'm checking that the first character starts with a digit, I know the result can not be negative. I think this will suffice then:

```
long long val = strtoll(source_date_epoch, &endptr, 10);
time_t create_time = val;
if (!isdigit(*source_date_epoch) || *endptr != '\0' || errno != 0 || (long long)create_time != val) {
    /* conversion failed */
}
```

(I'm not handling time_t as float - I'm not that pedantic.)